src/linalg.jl

```diff
@@ -149,3 +149,21 @@ function _mul!(C::BlockDiagonal, A::BlockDiagonal, B::BlockDiagonal, α::Number,
 
     return C
 end
+
+function LinearAlgebra.:\(B::BlockDiagonal, vm::AbstractVecOrMat)
+    row_i = 1
+    # BlockDiagonals with non-square blocks
+    all(b -> size(b, 1) == size(b, 2), blocks(B)) || throw(
```

Comment:
why should we error here rather than fallback to the `Matrix(B) \ vm`?
I think we should just fallback.

if we are going to error here, then that's a breaking change, because `B \ vm` currently accepts BlockDiagonals of all structures (since it fallsback to `\(A::AbstractMatrix, B::AbstractVecOrMat)`

Reply:
So the Base behavior is : 
>   For rectangular A the result is the minimum-norm least squares solution computed by a pivoted QR factorization of A and a rank estimate of `A` based on the `R` factor.

I guess the best solution would be to try to make it work anyway with rectangular blocks?

Reply:
But in the scope of this PR I could just fall back and one could specialize on rectangular blocks in another PR?

Reply:
yah, since this is an optimization, i think just fallback to the slow thing to we can do a faster thing is better (since the slow thing is what already happens before this PR)

also, btw, just noticed we have an `is_square` function, so the check can be written:
```jl
    all(is_square, blocks(B))
```

Reply:
and I kept searching for a `issquare` function :laughing:

Reply:
I _think_ the fallback is unnecessary here. The reason being that if the blocks are not square the determinant will be zero and the matrix will not be invertible, so the fallback will necessarily fail, meaning that the fallback call is redundant, and the error message is less clear.

But I can't prove/find a reference for this so fallback will have to do it